File: internal/dashboard/components/watcher_groups.go
package components

import (
	"sort"
	"strings"
)

// WatcherSpaceGroup is a WTG workspace with its repo checkouts.
type WatcherSpaceGroup struct {
	Space    string
	Watchers []WatcherInfo
}

// GroupWatchers splits watchers into local (non-WTG) repos and space-grouped worktrees.
func GroupWatchers(watchers []WatcherInfo) (local []WatcherInfo, spaces []WatcherSpaceGroup) {
	bySpace := map[string][]WatcherInfo{}
	for _, w := range watchers {
		if strings.TrimSpace(w.Workspace) == "" {
			local = append(local, w)
			continue
		}
		bySpace[w.Workspace] = append(bySpace[w.Workspace], w)
	}
	sortWatchers(local)
	spaceNames := make([]string, 0, len(bySpace))
	for name := range bySpace {
		spaceNames = append(spaceNames, name)
	}
	sort.Strings(spaceNames)
	for _, name := range spaceNames {
		items := bySpace[name]
		sortWatchers(items)
		spaces = append(spaces, WatcherSpaceGroup{Space: name, Watchers: items})
	}
	return local, spaces
}

func sortWatchers(ws []WatcherInfo) {
	sort.Slice(ws, func(i, j int) bool {
		li, lj := strings.ToLower(ws[i].Label), strings.ToLower(ws[j].Label)
		if li != lj {
			return li < lj
		}
		return ws[i].ProjectPath < ws[j].ProjectPath
	})
}

File: ui/src/lib/watcherGroups.ts
import type { WatcherInfo } from '../api/types'

export type WatcherSpaceGroup = { space: string; watchers: WatcherInfo[] }

export type WatcherDisplayEntry = {
  watcher: WatcherInfo
  groupKey: string
  groupLabel: string
}

export function groupWatchers(watchers: WatcherInfo[]): { local: WatcherInfo[]; spaces: WatcherSpaceGroup[] } {
  const local: WatcherInfo[] = []
  const bySpace = new Map<string, WatcherInfo[]>()
  for (const w of watchers) {
    const space = (w.Workspace || '').trim()
    if (!space) {
      local.push(w)
      continue
    }
    const list = bySpace.get(space) || []
    list.push(w)
    bySpace.set(space, list)
  }
  sortWatchers(local)
  const spaces = [...bySpace.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([space, items]) => ({ space, watchers: sortWatchers([...items]) }))
  return { local, spaces }
}

/** Flat space watchers with per-workspace group labels for paginated Spaces column. */
export function flattenSpaceWatchers(spaces: WatcherSpaceGroup[]): WatcherDisplayEntry[] {
  const out: WatcherDisplayEntry[] = []
  for (const g of spaces) {
    for (const w of g.watchers) {
      out.push({ watcher: w, groupKey: `space:${g.space}`, groupLabel: g.space })
    }
  }
  return out
}

function sortWatchers(ws: WatcherInfo[]): WatcherInfo[] {
  ws.sort((a, b) => {
    const la = (a.Label || a.Name || '').toLowerCase()
    const lb = (b.Label || b.Name || '').toLowerCase()
    if (la !== lb) return la < lb ? -1 : 1
    return a.ProjectPath < b.ProjectPath ? -1 : a.ProjectPath > b.ProjectPath ? 1 : 0
  })
  return ws
}
